dcap-ql/src/quote.rs

```diff
@@ -0,0 +1,382 @@
+            qe3_report: qe3_report.into(),
+            qe3_signature: qe3_signature.into(),
+            authentication_data: authentication_data.into(),
+            certification_data_type,
+            certification_data: certification_data.into(),
+        })
+    }
+}
+
+impl<'a> Qe3CertData<'a> for Qe3CertDataPpid<'a> {
+    fn parse(type_: CertificationDataType, data: &'a [u8]) -> Result<Self> {
+        let ppid_len = match type_ {
+            CertificationDataType::PpidCleartext => bail!(
+                "PPID clear text not implemented. Data length = {}",
+                data.len()
+            ),
+            CertificationDataType::PpidEncryptedRsa2048 => 256,
+            CertificationDataType::PpidEncryptedRsa3072 => 384,
+            _ => bail!("Invalid certification data type: {:?}", type_),
+        };
+
+        let (ppid, rest) = data.try_split_at_map(ppid_len, Cow::from)?;
+        let (cpusvn, rest) = rest.try_split_at_map(CPUSVN_LEN, Cow::from)?;
+        let (pcesvn, rest) = rest.try_split_at_map(mem::size_of::<u16>(), LE::read_u16)?;
+        let (pceid, rest) = rest.try_split_at_map(mem::size_of::<u16>(), LE::read_u16)?;
+        if !rest.is_empty() {
+            bail!(
+                "Invalid certification data length for type {:?}: {}",
+                type_,
+                data.len()
+            );
+        }
+
+        Ok(Qe3CertDataPpid {
+            ppid: ppid.into(),
+            cpusvn: cpusvn.into(),
+            pcesvn,
+            pceid,
+        })
+    }
+}
+
+// =============================================
+// ================= ACCESSORS =================
+// =============================================
+
+impl<'a> Quote<'a> {
+    pub fn header(&self) -> &QuoteHeader<'a> {
+        &self.header
+    }
+
+    pub fn report_body(&self) -> &[u8] {
+        &self.report_body
+    }
+
+    pub fn signature<'b, T: Quote3Signature<'b>>(&'b self) -> Result<T> {
+        let QuoteHeader::V3 {
+            attestation_key_type,
+            ..
+        } = self.header;
+        T::parse(attestation_key_type, &self.signature)
+    }
+
+    pub fn clone_owned(&self) -> Quote<'static> {
+        Quote {
+            header: self.header.clone_owned(),
+            report_body: self.report_body.clone().into_owned().into(),
+            signature: self.signature.clone().into_owned().into(),
+        }
+    }
+}
+
+impl<'a> QuoteHeader<'a> {
+    pub fn clone_owned(&self) -> QuoteHeader<'static> {
+        match *self {
+            QuoteHeader::V3 {
+                attestation_key_type,
+                qe3_svn,
+                pce_svn,
+                ref qe3_vendor_id,
+                ref user_data,
+            } => QuoteHeader::V3 {
+                attestation_key_type,
+                qe3_svn,
+                pce_svn,
+                qe3_vendor_id: qe3_vendor_id.clone().into_owned().into(),
+                user_data: user_data.clone().into_owned().into(),
```

Comment:
I believe `.clone().into_owned()` on `Cow` is equivalent to just `.to_owned()`.

Reply:
`(**user_data).to_owned()` but yes